Expose and document ProjectSettings.get_global_class_list()

bitsawer opened the pull request:
Exposes a better way to get global script classes now that https://github.com/godotengine/godot/pull/70557 is merged. The currently supported and documented way feels really sketchy, you have to read and parse a random config file in a hidden directory. If I saw that in a project code without knowing it, I would assume it's some kind of fragile hack to get some undocumented data. We have the API call now, so why not expose it to users instead? This also makes it possible to easily change the actual storage implementation later without breaking all user code. Also see https://github.com/godotengine/godot-docs/issues/6632

akien-mga commented:
I think it's a good idea but this is not the right method to expose for this. You shouldn't re-read the config file every time a user calls this method.

Instead, it's probably the global class information from `ScriptServer` which should be exposed somehow.

bitsawer commented:
> I think it's a good idea but this is not the right method to expose for this. You shouldn't re-read the config file every time a user calls this method.
> 
> Instead, it's probably the global class information from `ScriptServer` which should be exposed somehow.

I was thinking of something similar (in minimum rename it to something like load_global_class_list()), but I'm not yet familiar with the Godot core to have much idea which mechanism should be used or exposed. Especially if it involves class cache invalidation etc. The PR is here at least as a discussion starter, anyone more familar with this are is welcome to pick it up or give more in-depth feedback. I think at least reading the config file directly as a documented way to get this data should be discouraged.

Structed commented:
It would be so cool if we could use `ClassDB`, as it exposes the API I really would want to use from an addon, for example. Here's some example code I tried to use, before discovering that custom classes are not exposed there:

```gdscript
func get_class_instance(name: String) -> RefCounted:
	var class_list = ClassDB.get_class_list()
	var has = class_list.has(name) # false
	var exists = ClassDB.class_exists(name) # false
	if ClassDB.can_instantiate(name):
		return ClassDB.instantiate(name)

	push_error("Class \"" + name + "\" could not be found")
	return null
```

bitsawer commented:
Updated the PR, it's ready for a review, @KoBeWi or anyone else. In the end, I simply ended storing/caching the class data inside ProjectSettings because that data is actually combined often with ScriptServer and EditorData when needed (for example class icons), there is no single place where all info is available. Now it is read from the config file when first needed and then maintained in memory (and also updated when store_global_class_list() is called). It might be possible to refactor things so that there is a single source of truth for all global class data, but I think that is outside the scope of this PR.

I tested it quite a bit including renaming, deleting and creating class_name-classes and everything seems to work, including deleting .godot-directory and re-importing the project.

MikeSchulze commented:
please commit it to the next beta ;)

akien-mga reviewed (commented):
Seems like I had a pending review from last week that I forgot to submit, here it is (some of the code was changed but it's still mostly relevant).

bitsawer committed:
Expose and document ProjectSettings.get_global_class_list()

bitsawer commented:
@akien-mga Updated the PR and resolved conflicts. The only suggestion I didn't do was adding `get_global_class_name(String path)`. Adding it would be pretty trivial, but I think people who use this API can now easily implement those themselves. And I'm sure next people would want `get_global_class_icon(String path)` and others.

I'm slightly against adding additional class-related API's to ProjectSettings, I think the original reason this data was even stored in ProjectSettings was mostly just historical and convenience, not any intentional API design policy. In hypothetical Godot 5.0 I would probably move all that stuff to ScriptServer (which has some other useful APIs that could be exposed) or even ClassDB (which could then offer API to non-engine classes too).

KoBeWi commented on core/config/project_settings.cpp:
```diff
@@ -1136,20 +1136,28 @@ Variant ProjectSettings::get_setting(const String &p_setting, const Variant &p_d
-Array ProjectSettings::get_global_class_list() {
-	Array script_classes;
+TypedArray<Dictionary> ProjectSettings::get_global_class_list() {
+	static bool loaded = false;
```
Why static local variable instead of a field in class?
Also right now the list can't be reloaded (which is important in editor). You could probably set the array in `store_global_class_list()`.

bitsawer replied:
I looked for some code examples about how that once-only init is done elsewhere in the engine source code and that pattern is used often. I guess it's a bit less mental burden as that static variable is only used inside one function, but I can easily change it if needed.

KoBeWi replied:
Any example? I only found it in print methods, but that's not the same use.

bitsawer replied:
In `Object::initialize_class()` there is `static bool initialized`, but now that I look it's not really that common. I I think that was just the first example I found and assumed it's more common as that is pretty core code. I could just change that to non-static member variable like "bool global_class_list_loaded" (free to bikeshed the name)?

KoBeWi replied:
Sound ok.

KoBeWi commented on core/config/project_settings.h:
```diff
@@ -36,6 +36,7 @@
 #include "core/templates/hash_map.h"
 #include "core/templates/local_vector.h"
 #include "core/templates/rb_set.h"
+#include "core/variant/typed_array.h"
```
Use a forward-declare instead and move the include to cpp:
```C++
template <typename T>
class TypedArray;
```

bitsawer replied:
I don't think we can have forward-declared class template members (as the compiler has to know the class size here). Returning or passing them as arguments seems to work fine. I could either:

* Keep the include and code as-is, or
* Use the forward declaration and change `TypedArray<Dictionary> global_class_list;` member variable into normal `Array global_class_list;` and then create and return TypedArray<Dictionary> from the get_global_class_list() only when it's called.

KoBeWi replied:
I think it's fine if the member is Array. Converting it is not expensive and the function is not called often.

YuriSizov commented on doc/classes/ProjectSettings.xml:
```diff
@@ -60,6 +60,17 @@
+		<method name="get_global_class_list">
+			<return type="Dictionary[]" />
+			<description>
+				Returns an Array of Dictionaries containing registered global class metadata.
+				[code]base[/code] - Name of the base class.
+				[code]class[/code] - Name of the class.
+				[code]icon[/code] - Path to the class icon.
+				[code]language[/code] - Programming language used to code the class.
+				[code]path[/code] - Path to the file containing the class.
```
```suggestion
				Returns an [Array] of registered global classes. Each global class is represented as a [Dictionary] that contains the following entries:
				- [code]base[/code] is a name of the base class;
				- [code]class[/code] is a name of the registered global class;
				- [code]icon[/code] is a path to a custom icon of the global class, if it has any;
				- [code]language[/code] is a name of a programming language in which the global class is written;
				- [code]path[/code] is a path to a file containing the global class.
				[b]Note:[/b] Both the script and the icon paths are local to the project filesystem, i.e. they start with [code]res://[/code].
```

YuriSizov commented:
Thanks!
